### PR TITLE
recover inactive state by manual activating producer

### PR DESCRIPTION
**Pull request by raindust**

### CODE REVIEW

File: core/types/payload/cancelproducer.go [hunk 1]
```diff
@@ -9,63 +9,84 @@ import (
 	"github.com/elastos/Elastos.ELA/crypto"
 )
 
-const CancelProducerVersion byte = 0x00
+type ProducerOperation byte
 
-type CancelProducer struct {
+const (
+	ProducerOperationVersion byte = 0x00
```
Comment (alexpan0610) [inline core/types/payload/cancelproducer.go]: Is this the ProcessProducerVersion, the version of the ProcessProducer payload?

File: core/types/payload/cancelproducer.go [hunk 2]
```diff
@@ -9,63 +9,84 @@ import (
...
+
+	OperationCancel   ProducerOperation = 0x01
```
Comment (alexpan0610) [inline core/types/payload/cancelproducer.go]: Flag definition often use short name as prefix. For example ServiceFlag -> SF,  SFNodeNetwork, SFNodeBloom.
So OperationCancel can be POCancel.

File: cmd/script/api/payloadtype.go [hunk 3]
```diff
@@ -304,12 +306,13 @@ func newCancelProducer(L *lua.LState) int {
 		fmt.Println("wrong producer public key")
 		os.Exit(1)
 	}
-	cancelProducer := &payload.CancelProducer{
+	cancelProducer := &payload.ProcessProducer{
```
Comment (alexpan0610) [inline cmd/script/api/payloadtype.go]: Update variable cancelProducer  to processProducer

File: blockchain/txvalidator.go [hunk 4]
```diff
@@ -775,29 +786,82 @@ func (b *BlockChain) checkRegisterProducerTransaction(txn *Transaction) error {
 	return nil
 }
 
-func (b *BlockChain) checkCancelProducerTransaction(txn *Transaction) error {
-	cancelProducer, ok := txn.Payload.(*payload.CancelProducer)
+func (b *BlockChain) checkProcessProducer(txn *Transaction) (
+	*state.Producer, error) {
+	cancelProducer, ok := txn.Payload.(*payload.ProcessProducer)
```
Comment (alexpan0610) [inline blockchain/txvalidator.go]: Update variable cancelProducer  to processProducer

File: core/types/payload/cancelproducer.go [hunk 5]
```diff
@@ -9,63 +9,84 @@ import (
 	"github.com/elastos/Elastos.ELA/crypto"
 )
 
-const CancelProducerVersion byte = 0x00
+type ProducerOperation byte
```
Comment (alexpan0610) [inline core/types/payload/cancelproducer.go]: Add comments to this flag definition.

File: dpos/state/state.go [hunk 6]
```diff
@@ -802,98 +831,86 @@ func (s *State) ProcessSpecialTxPayload(p types.Payload) {
...
+func (s *State) revertSettingInactiveProducer(producer *Producer, key string,
+	height uint32) {
+	producer.inactiveSince = 0
+	producer.state = Activate
+	s.activityProducers[key] = producer
+	delete(s.inactiveProducers, key)
 
-	for k, v := range s.inactiveProducers {
-		tryLeaveInactive(k, v)
+	if producer.penalty < s.chainParams.InactivePenalty {
```
Comment (alexpan0610) [inline dpos/state/state.go]: No if else in history func

File: dpos/state/state.go [hunk 7]
```diff
@@ -802,98 +831,86 @@ func (s *State) ProcessSpecialTxPayload(p types.Payload) {
...
 
-			if producer.state == Activate {
+	isMissOnDuty := !bytes.Equal(onDutyArbiter, confirm.Proposal.Sponsor)
+	if producer, ok := s.activityProducers[key]; ok {
+		count, existMissingRecord := s.onDutyMissingCounts[key]
 
-				if producer.inactiveRounds == 0 {
-					producer.inactiveSince = height
+		s.history.append(height, func() {
+			if isMissOnDuty {
```
Comment (alexpan0610) [inline dpos/state/state.go]: No if/else in history func.

File: dpos/state/state.go (hunk 6 above)
Comment (raindust) [reply dpos/state/state.go]: it's just a guard condition to prevent the penalty from becoming negative

**Comment by raindust** [comment]
Append the arrangement as a commit, squash me after finally approved

**Pull request merged**